### 1. Подберите скорость обучения (alpha) и количество итераций:

In [1]:
import numpy as np
import matplotlib.pyplot as plt

In [2]:
X = np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
              [1, 1, 2, 1, 3, 0, 5, 10, 1, 2]])
X.shape

(2, 10)

In [3]:
y = [45, 55, 50, 59, 65, 35, 75, 80, 50, 60]

In [4]:
def calc_mse(y, y_pred):
    err = np.mean((y - y_pred)**2)
    return err

In [5]:
alpha = 1e-4
iterations = 1001

n = X.shape[1]
W = np.array([1, 0.5])
W, alpha

(array([1. , 0.5]), 0.0001)

In [6]:
for i in range(iterations):
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)
    for ii in range(W.shape[0]):
        W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
        if i % 100 == 0:
            print(i, W, err, sep='\t')

0	[1.01102 0.5    ]	3173.15
0	[1.01102 0.534  ]	3173.15
100	[2.02168794 3.42791095]	2207.6942101483005
100	[2.02168794 3.452835  ]	2207.6942101483005
200	[2.88306663 5.57141207]	1670.7027004086372
200	[2.88306663 5.58962851]	1670.7027004086372
300	[3.63289519 7.1352209 ]	1368.0832412796256
300	[3.63289519 7.14848061]	1368.0832412796256
400	[4.29908617 8.27068189]	1193.7493344933293
400	[4.29908617 8.28027926]	1193.7493344933293
500	[4.90228732 9.08968104]	1089.714750358689
500	[4.90228732 9.09657298]	1089.714750358689
600	[5.45777535 9.6749329 ]	1024.2828984889416
600	[5.45777535 9.67982684]	1024.2828984889416
700	[ 5.97685612 10.08758579]	980.1260644155933
700	[ 5.97685612 10.0910047 ]	980.1260644155933
800	[ 6.46789973 10.37284419]	947.770026039189
800	[ 6.46789973 10.37517467]	947.770026039189
900	[ 6.93710574 10.56412541]	922.0330206910178
900	[ 6.93710574 10.56565326]	922.0330206910178
1000	[ 7.38906894 10.6861326 ]	900.0794345405571
1000	[ 7.38906894 10.68706909]	900.079434540557

In [7]:
alpha = 1e-5
iterations = 1000001

n = X.shape[1]
W = np.array([1, 0.5])
W, alpha

for i in range(iterations):
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)
    for ii in range(W.shape[0]):
        W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
        if i % 10000 == 0:
            print(i, W, err, sep='\t')

0	[1.001102 0.5     ]	3173.15
0	[1.001102 0.5034  ]	3173.15
10000	[ 7.38444627 10.68385176]	900.1284957199632
10000	[ 7.38444627 10.68394611]	900.1284957199632
20000	[11.40864377 10.49721082]	735.7038137817291
20000	[11.40864377 10.4971504 ]	735.7038137817291
30000	[14.95076689  9.87037593]	606.1848763872531
30000	[14.95076689  9.87031436]	606.1848763872531
40000	[18.13894003  9.28298388]	500.99384806664955
40000	[18.13894003  9.28292804]	500.99384806664955
50000	[21.01206086  8.75249998]	415.5533147207666
50000	[21.01206086  8.75244964]	415.5533147207666
60000	[23.6014356   8.27435036]	346.15493701121403
60000	[23.6014356   8.27430498]	346.15493701121403
70000	[25.93509516  7.84341794]	289.7866695863051
70000	[25.93509516  7.84337705]	289.7866695863051
80000	[28.03829326  7.45504229]	244.00200318110146
80000	[28.03829326  7.45500544]	244.00200318110146
90000	[29.93378922  7.10502083]	206.81378055872133
90000	[29.93378922  7.10498761]	206.81378055872133
100000	[31.64209475  6.78956586]

810000	[47.22244495  3.9125051 ]	45.93755056366078
810000	[47.22244495  3.91250508]	45.93755056366078
820000	[47.22340267  3.91232824]	45.93754106992191
820000	[47.22340267  3.91232823]	45.93754106992191
830000	[47.22426581  3.91216886]	45.937533358709786
830000	[47.22426581  3.91216884]	45.937533358709786
840000	[47.22504371  3.91202521]	45.9375270953405
840000	[47.22504371  3.9120252 ]	45.9375270953405
850000	[47.22574479  3.91189575]	45.937522007969775
850000	[47.22574479  3.91189574]	45.937522007969775
860000	[47.22637663  3.91177907]	45.93751787579434
860000	[47.22637663  3.91177906]	45.93751787579434
870000	[47.22694608  3.91167392]	45.93751451946842
870000	[47.22694608  3.91167391]	45.93751451946842
880000	[47.22745929  3.91157915]	45.93751179332005
880000	[47.22745929  3.91157914]	45.93751179332005
890000	[47.22792182  3.91149374]	45.93750957902823
890000	[47.22792182  3.91149373]	45.93750957902823
900000	[47.22833867  3.91141677]	45.93750778048773
900000	[47.22833867  3.911416

In [8]:
alpha = 5e-3
iterations = 10001

n = X.shape[1]
W = np.array([1, 0.5])
W, alpha

for i in range(iterations):
    y_pred = np.dot(W, X)
    err = calc_mse(y, y_pred)
    for ii in range(W.shape[0]):
        W[ii] -= alpha * (1/n * 2 * np.sum(X[ii] * (y_pred - y)))
        if i % 1000 == 0:
            print(i, W, err, sep='\t')

0	[1.551 0.5  ]	3173.15
0	[1.551 2.2  ]	3173.15
1000	[46.99313195  3.95508013]	45.96853390140358
1000	[46.99313195  3.95484948]	45.96853390140358
2000	[47.23084067  3.910956  ]	45.937500921189624
2000	[47.23084067  3.91095474]	45.937500921189624
3000	[47.23213576  3.9107156 ]	45.937500000027356
3000	[47.23213576  3.9107156 ]	45.937500000027356
4000	[47.23214282  3.91071429]	45.9375
4000	[47.23214282  3.91071429]	45.9375
5000	[47.23214286  3.91071429]	45.93750000000001
5000	[47.23214286  3.91071429]	45.93750000000001
6000	[47.23214286  3.91071429]	45.93749999999999
6000	[47.23214286  3.91071429]	45.93749999999999
7000	[47.23214286  3.91071429]	45.93749999999998
7000	[47.23214286  3.91071429]	45.93749999999998
8000	[47.23214286  3.91071429]	45.93749999999998
8000	[47.23214286  3.91071429]	45.93749999999998
9000	[47.23214286  3.91071429]	45.93749999999998
9000	[47.23214286  3.91071429]	45.93749999999998
10000	[47.23214286  3.91071429]	45.93749999999998
10000	[47.23214286  3.91071429]	45.9

при альфа 0.005 изменения параметров отсутствуют уже после 5000-й итерации

`i=5000	W=[47.23214286  3.91071429]	err=45.9375`

при столь широком шаге есть небольшая вероятность перескочить минимум, но в данном случае этого не произошло